dnf/base.py

```diff
@@ -1744,9 +1753,111 @@ def install_module(self, specs, strict=True):
+        glob_exclude_query = self.sack.query().filter(name__glob=glob_excludes)
+        self.sack.add_excludes(exclude_query)
+        self.sack.add_excludes(glob_exclude_query)
+
+    def install_packages(self, package_specs, ignored):
+        for spec in package_specs:
+            try:
+                self.install(spec, strict=self.conf.strict)
+            except dnf.exceptions.Error:
+                ignored.append(spec)
+
+    def install_modules(self, module_specs):
+        # temporary hack, in future raise exception/add to ignored
+        # when enabling different stream for kickstart
+        former_assumeyes = self.conf.assumeyes
+        if not self.output:
+            self.conf.assumeyes = True
+        groups = self.install_module(module_specs)
+        self.conf.assumeyes = former_assumeyes
+        return groups
+
+    def exclude_groups(self, group_specs):
+        group_excludes = []
+
+        for group_spec in group_specs:
+            if '/' in group_spec:
+                split = group_spec.split('/')
+                group_spec = split[0]
+
+            if group_spec[0] == "^":
+                environment = self.comps.environment_by_pattern(group_spec[1:])
+                if not environment:
+                    continue
+
+            for group in environment.groups_iter():
+                for pkg in group.packages_iter():
+                    group_excludes.append(pkg.name)
+
+        exclude_query = self.sack.query().filter(name=group_excludes)
+        self.sack.add_excludes(exclude_query)
+
+    def install_groups(self, group_specs, exclude_packages, ignored):
+        for group_spec in group_specs:
+            try:
+                types = self.conf.group_package_types
+
+                if '/' in group_spec:
+                    split = group_spec.split('/')
+                    group_spec = split[0]
+                    types = split[1].split(',')
+
+                if group_spec[0] == "^":
+                    environment = self.comps.environment_by_pattern(group_spec[1:])
+                    if environment:
+                        self.environment_install(environment.id,
+                                                 types,
+                                                 exclude=exclude_packages,
+                                                 strict=self.conf.strict)
+                    else:
+                        ignored.append("@" + group_spec)
+                else:
+                    group = self.comps.group_by_pattern(group_spec)
+
+                if group:
+                    self.group_install(group.id, types,
+                                       exclude=exclude_packages,
+                                       strict=self.conf.strict)
+                else:
+                    ignored.append("@" + group_spec)
+            except dnf.exceptions.Error:
+                ignored.append("@" + group_spec)
+
+    def install_specs(self, install, exclude=None):
```

Comment:
Please add following arguments:
strict=True
reponame=None
forms=none
pkg_types

Reply:
Please can you also redesign the def to shape that will replace the logic in install.py run(). I think the redundancy is important to eliminate.

Reply:
I added the arguments, but refactoring cli.commands.install.run() is not goal of this PR. It can be done in future. This PR need to be done ASAP.